feat: STRUCT and ARRAY support

jimfulton opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #293 
Fixes #314 
Fixed #233 
Fixes #37 
🦕

Commit:
feat: STRUCT and ARRAY support

Commit:
Merge branch 'main' into struct

Commit:
Fixed test that expected JSON rather than STRUCT

Commit:
Merge branch 'struct' of github.com:jimfulton/python-bigquery-sqlalchemy into struct

Commit:
Added system test I neglected to check in before :(

Commit:
blacken

Commit:
Merge branch 'main' into struct

jimfulton commented:
FTR, WRT superset, once I finally got it working :), it behaves the same with and without these changes.

BTW, we have logic that tries to unpack sub-structs, I think so that there would eventually be scalars for superset to work with.

If you have an array of structs, we still create columns for the fields of the struct in the array.  This causes superset to error, because it has no way to get  at structs in an array.  We should probably not unpack structs in arrays.

Commit:
Don't strip <ARRAY > from parameter types

Otherwise, the BQ doesn't handle arrays of structs.

Commit:
Added system tests to verift PR 67 and issue 233

Commit:
Merge branch 'struct' of github.com:jimfulton/python-bigquery-sqlalchemy into struct

Commit:
blacken

Commit:
Renamed test file to conform to samples test-file naming conventions

Commit:
Require google-cloud-bigquery 2.25.2 to get struct field-name underscore fix

Commit:
Added STRUCT documentation

Commit:
fix bigquery version

Commit:
Merge branch 'main' into struct

Commit:
get blacken to leave sample code alone.

I want it narrow to avoid horizonal scrolling

Commit:
Check in missing file :(

Commit:
Merge branch 'struct' of github.com:jimfulton/python-bigquery-sqlalchemy into struct

Commit:
need sqla 1.4 for unnest

Commit:
fixed typo

Commit:
Merge branch 'main' into struct

jimfulton commented:
Some notes for reviewers:

- Heart of change is `_struct.py`, which isn't large, but also isn't obvious.  :(  I cribbed from the built-in JSON and ARRAY types. When reviewing, it's probably helpful to look at those. The "Comparator" framework is confusing, in large part because the name doesn't make sense.

  Having said that, the core logic is in `_setop_getitem`, which is a hook used by the `Indexible` base class of `STRUCT`s comparator.
  
  The `__getattr__` method just delegates to (the inherited) `__getitem__`.
  
- This PR also has 2 other small changes:
  - Machinery for mapping BQ types to SQLAlchemy types has been factored into a separate `_types` module, both to avoid cluttering `base` more and to partially avoid circular imports.  (There's still a circular import issue that isn't fixable without a bigger refactoring that I deemed unwarranted.)
  - Implementation of ARRAY indexing, which wasn't implemented.  A number of my tests (see `test__struct.py` in both unit and system tests) used an example that has nested structs and arrays.

Commit:
Merge branch 'main' into struct

Commit:
Merge branch 'main' into struct

Commit:
Merge branch 'main' into struct

tswast commented on docs/struct.rst:
```diff
@@ -0,0 +1,69 @@
+   :start-after: [START bigquery_sqlalchemy_query_getitem]
+   :end-before: [END bigquery_sqlalchemy_query_getitem]
+
+and field names are case insensitive:
```
Is this because of a client-side convenience? Or is it BigQuery API behavior? Is it possible to have two struct fields that only differ by case?

jimfulton replied:
> Is this because of a client-side convenience?

It is convenient. :)  But that's not the reason.

> Or is it BigQuery API behavior?

Yes and also SQL.

> Is it possible to have two struct fields that only differ by case?

no

```
>>> client.query("create table sample_struct (s struct<x INT64, X int64>)").result()
Traceback (most recent call last):
...
google.api_core.exceptions.BadRequest: 400 Field s.X already exists in schema

(job ID: 040b2a36-9157-47f4-a1f4-c7000dfb3e9c)

              -----Query Job SQL Follows-----               

    |    .    |    .    |    .    |    .    |    .    |
   1:create table sample_struct (s struct<x INT64, X int64>)
    |    .    |    .    |    .    |    .    |    .    |
```

tswast commented on setup.py:
```diff
@@ -83,7 +83,7 @@ def readme():
         # https://github.com/googleapis/google-cloud-python/issues/10566
         "google-auth>=1.25.0,<3.0.0dev",  # Work around pip wack.
-        "google-cloud-bigquery>=2.24.1",
+        "google-cloud-bigquery>=2.25.2,<3.0.0dev",
```
Presumably for https://github.com/googleapis/python-bigquery/pull/930 ?

I'm okay with this, but would like to consider the alternatives.

Is there logic that we'd have to change to support older versions of `google-cloud-bigquery`? Or can we disable a few failing tests when using Python 3.6 (which constrains to the older package)?

jimfulton replied:
> Is there logic that we'd have to change to support older versions of `google-cloud-bigquery`?

We could monkey-patch the 2 broken functions.

> Or can we disable a few failing tests when using Python 3.6 (which constrains to the older package)?

We could, but then folks might be surprised when valid structs weren't handled correctly.

tswast replied:
OK, let's leave this as you've done it. I like the stricter requirements to either of these options.

tswast commented on sqlalchemy_bigquery/_struct.py:
```diff
@@ -0,0 +1,124 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
Because of the history of this package, it's not Apache 2.0 like our `google-cloud-*` client libraries are.

```suggestion
# Copyright (c) 2021 The sqlalchemy-bigquery Authors
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
# the Software, and to permit persons to whom the Software is furnished to do so,
# subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```

jimfulton replied:
Sorry. Bad copy-paste

tswast commented on sqlalchemy_bigquery/_struct.py:
```diff
@@ -0,0 +1,124 @@
+type_compiler = None
+
+
+class STRUCT(sqlalchemy.sql.sqltypes.Indexable, sqlalchemy.types.UserDefinedType):
```
Should we have a docstring on this?

tswast replied:
Looks like there's not much in other SQLAlchemy types, e.g. https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.Integer

Something like "A type for BigQuery STRUCT/RECORD data." would be in line with the other docstrings I'm seeing.

jimfulton replied:
sure

jimfulton replied:
Done. I included a doc link.

I also added a doc link for GEOGRAPHY.

tswast commented on sqlalchemy_bigquery/_struct.py:
```diff
@@ -0,0 +1,124 @@
+            for (name, type_) in (fields + tuple(kwfields.items()))
+        )
+
+        self.__byname = {name.lower(): type_ for (name, type_) in self.__fields}
```
Let's check the BigQuery behavior when two fields differ only by case. If similarly named fields are allowed, we should keep the original casing, even if it would be nice to have case insensitivity most of the time.

tswast replied:
Looks like you're right.

```
CREATE TABLE my_dataset.struct_test
(
    struct_col STRUCT<Field INT64, field INT64>
)
```

Fails with "struct_col.field already exists"

Querying a table with different cases works fine:

```
CREATE TABLE my_dataset.struct_test
(
    struct_col STRUCT<field1 INT64, field2 STRING>
)
AS SELECT struct_col from UNNEST(
    [
        (STRUCT<struct_col STRUCT<field1 INT64, field2 STRING>>((123, "Hello"))),
        (STRUCT<struct_col STRUCT<field1 INT64, field2 STRING>>((456, "World")))
    ]
)
```

```
SELECT struct_col.Field1 as capped, struct_col.field1 as lowered FROM my_dataset.struct_test
```

Perhaps we'll hear from folks on https://issuetracker.google.com/199114438 what the desired behavior is, but this looks good to me, since they can't change from case insensitive to case sensitive without breaking existing queries.

jimfulton replied:
Oh look, case sensitivity is documented: https://cloud.google.com/bigquery/docs/reference/standard-sql/lexical#case_sensitivity

tswast replied:
Good find!

tswast commented on sqlalchemy_bigquery/_struct.py:
```diff
@@ -0,0 +1,124 @@
+        *fields: Tuple[str, sqlalchemy.types.TypeEngine],
+        **kwfields: Mapping[str, sqlalchemy.types.TypeEngine],
+    ):
+        self.__fields = tuple(
```
Are we relying on one or the other superclasses to do something with these fields? Let's comment if so.

Also, should we be `__init__`ing one or both of our super classes?

jimfulton replied:
> Are we relying on one or the other superclasses to do something with these fields? Let's comment if so.

No.

> Also, should we be `__init__`ing one or both of our super classes?

No.

See: https://docs.sqlalchemy.org/en/14/core/custom_types.html#creating-new-types

and also: [ARRAY](https://github.com/sqlalchemy/sqlalchemy/blob/06f031e55f40403996e3580bdc5a22661b6303f4/lib/sqlalchemy/sql/sqltypes.py#L2661), which subclasses  `Indexable` and other things but doesn't call superclass `__init__`s.

We could call `super().__init__`, which would feel hygienic :), but would be a noop and might make the reader wonder what it's doing.  I would be OK either way.

jimfulton replied:
`__fields` is completely specific to `STRUCT` (this the leading `__`).

tswast reviewed (changes requested):
I haven't quite digested everything in this PR yet, but I figured I'd share the feedback I have so far.

jimfulton committed:
Update sqlalchemy_bigquery/_struct.py

Co-authored-by: Tim Swast <swast@google.com>

Commit:
added STRUCT docstring

Commit:
Add doc link

Commit:
Merge branch 'struct' of github.com:jimfulton/python-bigquery-sqlalchemy into struct

Commit:
Added some comments

tswast commented on sqlalchemy_bigquery/_types.py:
```diff
@@ -0,0 +1,141 @@
+                for f in field.fields
+            ]
+            results += _get_transitive_schema_fields(sub_fields, cur_fields)
+            cur_fields.pop()
```
Since we pop these off, does that mean we don't get the top-level struct field, just the leaf fields? I suspect this might hide some ARRAY columns if a parent node has mode `REPEATED`, but is not included.

Edit: I see the top field is added to results on line 83. Might be worth a comment as to why we pop here.

jimfulton replied:
I got rid of cur_fields. It wasn't needed (anymore).

tswast commented on sqlalchemy_bigquery/_types.py:
```diff
@@ -0,0 +1,141 @@
+    for field in fields:
+        results += [field]
+        if field.field_type == "RECORD":
+            cur_fields.append(field)
```
I don't quite understand what `cur_fields` is doing. Is there a better name we can pick for this? Maybe it's referring to `ancestors`?

jimfulton replied:
Haha, it's not doing anything.

This is based on https://github.com/googleapis/python-bigquery-sqlalchemy/blob/db644243040a784f10d3d5f967963b2acaeb15fc/pybigquery/sqlalchemy_bigquery.py#L503-L523, which I inherited.

I've refactored it quite a bit and failed to notice that this wasn't needed any more. Fixed.

tswast commented on tests/unit/test__struct.py:
```diff
@@ -0,0 +1,150 @@
+        (_col().NAME, "`t`.`person`.NAME"),
+        (_col().children, "`t`.`person`.children"),
+        (
+            _col().children[0].label("anon_1"),  # SQLAlchemy doesn't add the label
```
Should we file an issue for this to investigate later? If so, let's add `TODO` and link to the issue.

jimfulton replied:
Done

tswast reviewed (changes requested):
I like it! Just a few things I think we should clarify before merging.

Commit:
Localize logic for getting subtye column specifications

Commit:
explain semi-private name mangling

Commit:
Make name magling more explicit

Commit:
explain why we have different implementations of _field_index for SQLAlchemy 1.3 and 1/4

Commit:
get rid of cur_fields, we're not using it anymore.

Also, check for both RECORD and STRUCT fild types, in case the API
ever starts returning STRUCT.

Commit:
Add a todo to find out why Sqlalchemy doesn't generate an alias when accessing array items

jimfulton committed:
user `repr` rather than `str` to shpow an object in an error message

Co-authored-by: Tim Swast <swast@google.com>

jimfulton commented on sqlalchemy_bigquery/_struct.py:
```diff
@@ -0,0 +1,140 @@
+            process = type_compiler.process
+        except AttributeError:
+            type_compiler = base.dialect.type_compiler(base.dialect())
+            process = type_compiler.process
```
I refactored so this is combined and isolated in one place using a new, better named `_get_subtype_col_spec` function.

jimfulton commented on sqlalchemy_bigquery/_struct.py:
```diff
@@ -0,0 +1,140 @@
+            type_compiler = base.dialect.type_compiler(base.dialect())
+            process = type_compiler.process
+
+        fields = ", ".join(f"{name} {process(type_)}" for name, type_ in self.__fields)
```
yes

jimfulton commented on sqlalchemy_bigquery/_struct.py:
```diff
@@ -0,0 +1,140 @@
+                    f"STRUCT fields can only be accessed with strings field names,"
+                    f" not {name}."
+                )
+            subtype = self.expr.type._STRUCT__byname.get(name.lower())
```
I refactored this to make name mangling more explicit and consistent, so I don't think comments are needed anymore. See if you agree. :)

I mainly use "private" variables, which aren't :), to avoid namespace conflicts when subclassing across responsibility boundaries. Arguably, explicit naming is better.

jimfulton commented on sqlalchemy_bigquery/_struct.py:
```diff
@@ -0,0 +1,140 @@
+            return operator, index, subtype
+
+        def __getattr__(self, name):
+            if name.lower() in self.expr.type._STRUCT__byname:
```
See my response on name mangling

jimfulton commented on sqlalchemy_bigquery/_struct.py:
```diff
@@ -0,0 +1,140 @@
+
+
+if sqlalchemy_1_4_or_more:
+
```
I added one that tried to explain.  I can't actually explain what this code is doing. It's stolen and I have to steel different code for different versions of SQLAlchemy.  SQLAlchemy is a complicated beast.

tswast commented on sqlalchemy_bigquery/_struct.py:
```diff
@@ -34,8 +34,13 @@
+    global _get_subtype_col_spec
+
+    type_compiler = base.dialect.type_compiler(base.dialect())
+    _get_subtype_col_spec = type_compiler.process
```
Fancy! I didn't realize a function could replace itself. I like it.

AnastasiaTamazlykar commented on sqlalchemy_bigquery/base.py:
```diff
@@ -566,6 +514,11 @@ def visit_bindparam(
+    def visit_getitem_binary(self, binary, operator_, **kw):
+        left = self.process(binary.left, **kw)
+        right = self.process(binary.right, **kw)
+        return f"{left}[OFFSET({right})]"
```
SAFE_OFFSET support would be nice too
